changes/2024-05-20-keystore-kms-config/change.md

```diff
@@ -0,0 +1,274 @@
+[//]: # "SPDX-License-Identifier: CC-BY-SA-4.0"
+
+# Branch Keystore KMS Configuration of Discovery
+
+## Affected Features
+
+This serves as a reference of all features that this change affects.
+
+| Feature                                         |
+| ----------------------------------------------- |
+| [Keystore](../../framework/branch-key-store.md) |
+
+## Affected Specifications
+
+This serves as a reference of all specification documents that this change affects.
+
+| Specification                                                                           |
+| --------------------------------------------------------------------------------------- |
+| [Keystore](../../framework/branch-key-store.md)                                         |
+| [AWS KMS Hierarchical Keyring](../../framework/aws-kms/aws-kms-hierarchical-keyring.md) |
+
+## Affected Implementations
+
+| Language | Version Introduced | Version Removed | Implementation                                                                                                                                                                                                                   |
+| -------- | ------------------ | --------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| Dafny    | 1.4.0              | n/a             | [AwsCryptographyKeyStoreOperations.dfy](https://github.com/aws/aws-cryptographic-material-providers-library/blob/main/AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStore/src/AwsCryptographyKeyStoreOperations.dfy) |
+| Java     | 1.4.0              | n/a             | [KeyStore.java](https://github.com/aws/aws-cryptographic-material-providers-library/blob/main/AwsCryptographicMaterialProviders/runtimes/java/src/main/smithy-generated/software/amazon/cryptography/keystore/KeyStore.java)     |
+| .NET     | 1.4.0              | n/a             | [KeyStore.cs](https://github.com/aws/aws-cryptographic-material-providers-library/blob/main/AwsCryptographicMaterialProviders/runtimes/net/Generated/AwsCryptographyKeyStore/KeyStore.cs)                                        |
+
+## Definitions
+
+An "MPL Consumer" is a library, service, or other application
+that uses the AWS Cryptographic Material Providers Library (MPL)
+to manage cryptographic materials.
+An "MPL Consumer" may or may not be using an AWS Crypto Tools product,
+such as the AWS Encryption SDK or AWS Database Encryption SDK.
+
+A "discovery Keystore" is a (Branch) Keystore with
+a KMS Configuration of Discovery or Multi-Region Discovery (MRDiscovery).
+
+A "static Keystore" is a (Branch) Keystore with
+a KMS Configuration of KMS Key ARN.
+
+A "strict Keystore" is a (Branch) Keystore with
+a KMS Configuration of KMS Multi-Region Key ARN or KMS Key ARN.
```

Comment:
Why do we have both? What is the difference?

Reply:
1. The difference is clearly described. Please, compare line 43 to line 46.
2. I want a short hand for "a Keystore with a KMS Configuration of KMS Key ARN". That is 10 words! "Static Keystore" is 2!
3. I need to describe the "change". We started with only the "static Keystore". Today, we have "strict Keystore", tomorrow, we will have "Discovery Keystore".

Reply:
Andy's recent change, https://github.com/awslabs/aws-encryption-sdk-specification/commit/82911ea396c72a71098eead11d3aca3bf058acd0, is an example of a "static" behavior.

While VersionKey and CreateKey are supported with Strict KMS Configurations,
the KMS-ARN is always static,
even if a MKR Configuration is used when invoking VerisonKey.

I believe it is valuable to introduce this language.

I grant you, there is functionally no difference between a "static Keystore" and a "strict Keystore",
but we did NOT start there.

At one time, we were going to only allow VersionKey & CreateKey from "static Keystores".

Unless anyone strongly objects, 
I am inclined to leave this here.